library/std/src/sys/wasm/mutex_atomics.rs

```diff
@@ -28,11 +28,14 @@ impl Mutex {
 
     pub unsafe fn lock(&self) {
         while !self.try_lock() {
-            let val = wasm32::memory_atomic_wait32(
-                self.ptr(),
-                1,  // we expect our mutex is locked
-                -1, // wait infinitely
-            );
+            // SAFETY: the caller must uphold the safety contract for `memory_atomic_wait32`.
```

Comment:
I am uncertain what the preconditions we ask for callers to meet on lock() here -- cc @m-ou-se -- you've recently worked here, it would be good to write up some docs somewhere on those preconditions or make lock safe...

However, AFAICT, those preconditions are not necessary; this is locally safe by asserting that `ptr()` returns a valid memory address (which it does), so our comment should say something like "SAFETY: Passed pointer is valid for atomic 32-bit operations."

There are several other such cases -- please try to fix them as well.

Reply:
Yup, this `memory_atomic_wait32` function only requires the pointer to point to a valid 32-bit value (like i32 or AtomicU32) that it will read, nothing else. There's now safe wrappers in `sys/wasm/futex_atomics.rs` that make this guarantee by requiring a `&AtomicI32`. The unsafety in wasm's condvar/mutex/etc. could be avoided by using `AtomicI32` and those wrappers instead. But that can be a separate PR.

> make lock safe

For some platforms (wasm, cloudabi, windows), the mutex implementations could be safe. But some others (unix) have extra requirements about pinning/moving. That's why all of the locks on all platforms are unsafe, to not make a commitment about them being safe. Maybe the unsafe ones can be replaced at some point by an alternative implementation (e.g. parking_lot or some other implementation), but that still requires a lot of work.

Reply:
Makes sense. It would be great to add simple comments that the unsafety is for uniformity rather than any actual unsafety, to avoid confusion over any preconditions we need to check, but that should definitely be a separate PR.